python/lsst/daf/butler/_butler_repo_index.py

```diff
@@ -136,16 +129,29 @@ def _read_repository_index_from_environment(cls) -> Config:
         """
         cls._most_recent_failure = ""
         try:
-            index_uri = cls._get_index_uri()
-        except KeyError as e:
-            cls._most_recent_failure = str(e)
-            raise
-        try:
-            repo_index = cls._read_repository_index(index_uri)
+            index_uri = os.getenv(cls.index_env_var)
+            direct_configuration = os.getenv(cls.repositories_env_var)
+
+            if index_uri and direct_configuration:
+                raise RuntimeError(
+                    f"Only one of the environment variables {cls.repositories_env_var} and"
+                    f" {cls.index_env_var} should be set."
+                )
+
+            if direct_configuration:
+                return cls._validate_configuration(yaml.safe_load(direct_configuration))
```

Comment:
Have I missed where this is cached or is it parsing the YAML every single time it's called? `_read_repository_index` has cached the result of reading the external file. YAML parsing is notoriously slow.

Reply:
I didn't think caching was worth the added complexity to avoid parsing a few hundred bytes of in-memory data.  For instantiations via Butler() we are going to go off and read a file from the network, then do dozens of database round trips immediately after this.  LabeledButlerFactory only reads each label once.